File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-2.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n=s.length();

        Map<Character, Integer> mp=new HashMap<>();
        int res=0;

        int count=0;
        int i=0,j=0;
        while(i<n && j<n)
        {
            if(j<n &&!mp.containsKey(s.charAt(j)))
            {
                count++;
                mp.put(s.charAt(j),j);
            }
            else
            {
                int index=mp.get(s.charAt(j));
                res=Math.max(res,count);
                while(i<n && i<=index)
                {
                    mp.remove(s.charAt(i));
                    i++;
                    count--;
                }
                count++;
                mp.put(s.charAt(j),j);

            }
            j++;
        }
        res=Math.max(res,count);
        return res;
    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-2.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:

        res = 0
        count = 0
        n = len(grid)
        m = len(grid[0])

        dr = [[0, 1], [0, -1], [1, 0], [-1, 0]]

        def dfs(i, j):

            nonlocal n, m, count

            if i >= n or j >= m or i < 0 or j < 0 or grid[i][j] == 0:
                return

            grid[i][j] = 0
            count += 1

            for di, dj in dr:

                nr = di + i
                nc = dj + j
                dfs(nr, nc)

        for i in range(n):
            for j in range(m):
                count = 0
                if grid[i][j] == 1:
                    dfs(i, j)
                    res = max(res, count)

        return res
